### PR TITLE
SCT-12 Creates the SCSt sink that can launch tasks.

### PR DESCRIPTION
**Pull request by cppwfs**
resolves  spring-cloud/spring-cloud-task#12

### CODE REVIEW

File: spring-cloud-task-docs/src/main/asciidoc/features.adoc [hunk 1]
```diff
@@ -212,3 +212,39 @@ marking the final state of the task.
 }
 ```
 
+=== Launching a task from a Spring cloud stream
```
Comment (mminella) [inline spring-cloud-task-docs/src/main/asciidoc/features.adoc]: Should be `Spring Cloud Stream` (caps Cloud and Stream)

File: spring-cloud-task-docs/src/main/asciidoc/features.adoc [hunk 2]
```diff
@@ -212,3 +212,39 @@ marking the final state of the task.
...
+
+Allows a user to launch tasks from a stream.  This is done by creating a sink that
+listens for a message that contains a `TaskLaunchRequest` as a part of its payload.
```
Comment (mminella) [inline spring-cloud-task-docs/src/main/asciidoc/features.adoc]: "as a part of its payload" or "as the payload"?  If you got something other than a `TaskLaunchRequest`...would it work?

File: spring-cloud-task-docs/src/main/asciidoc/features.adoc [hunk 3]
```diff
@@ -212,3 +212,39 @@ marking the final state of the task.
...
+For example using Spring Cloud Dataflow a stream can be created that has a processor
```
Comment (mminella) [inline spring-cloud-task-docs/src/main/asciidoc/features.adoc]: `Spring Cloud Data Flow`

File: spring-cloud-task-docs/src/main/asciidoc/features.adoc [hunk 4]
```diff
@@ -212,3 +212,39 @@ marking the final state of the task.
...
+@SpringBootApplication
+@EnableTaskLauncher
+public class TaskSinkApplication {
+	public static void main(String[] args) {
+		SpringApplication.run(TaskSinkApplication.class, args);
+	}
+}
+```
+
+A sample TaskSink and TaskProcessor has been made available to you in the samples module
```
Comment (mminella) [inline spring-cloud-task-docs/src/main/asciidoc/features.adoc]: Is there something special about a `TaskSink` and `TaskProcessor`?  I think this should be just `Sink` and `Processor` respectively.

File: spring-cloud-task-docs/src/main/asciidoc/features.adoc [hunk 5]
```diff
@@ -212,3 +212,39 @@ marking the final state of the task.
...
+of the Spring Cloud Task project.  To install these samples into your local maven
+repository execute a maven build from the `spring-cloud-task-samples` directory with the
+property `skipInstall` set to false.  For example:
+`mvn clean install -DskipInstall=false`.  To register them with
+Spring Cloud Dataflow, execute the following commands from Spring Cloud Dataflow shell:
+```
+module register --name taskSink --type sink --coordinates io.spring:tasksink:1.0.0.BUILD-SNAPSHOT
+module register --name taskProcessor --type processor --coordinates io.spring:taskprocessor:1.0.0.BUILD-SNAPSHOT
+```
```
Comment (mminella) [inline spring-cloud-task-docs/src/main/asciidoc/features.adoc]: All of this should work outside of the scope of SCDF.  I'm thinking that our documentation should be SCDF agnostic.  I'm fine with including it as an example, but we need to send a clear message that you can do all of this without SCDF.

File: spring-cloud-task-docs/src/main/asciidoc/features.adoc (hunk 2 above)
Comment (cppwfs) [inline spring-cloud-task-docs/src/main/asciidoc/features.adoc]: It will throw an SpelEvaluationException because it does not handle other message types.  I 'll update the docs.

File: spring-cloud-task-samples/taskprocessor/README.adoc [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+* *data-source-url* sets the spring_datasource_url for the task.  Default is null.
+* *data-source-driver-class-name* establishes the spring_datasource_driverClassName for the task.  Default is null.
+* *data-source-user-name* establishes the spring_datasource_username for the task.  Default is null.
+* *data-source-password* establishes the spring_datasource_password for the task.  Default is null.
+
+A sample processor stream definition using all parameters:
+
+```
+stream create foo --definition "http --server.port=9000|taskProcessor --group=io.spring --artifact=mytask --data-source-url=jdbc:mysql://localhost:3306/practice --data-source-driver-class-name=com.mysql.jdbc.Driver --data-source-user-name=root --data-source-password=password --classifiers=exec|taskSink" --deploy
+```
```
Comment (mminella) [inline spring-cloud-task-samples/taskprocessor/README.adoc]: Same point as before about being SCDF agnostic in our documentation.

File: spring-cloud-task-samples/taskprocessor/README.adoc [hunk 7]
```diff
@@ -0,0 +1,44 @@
+= TaskProcessor
```
Comment (mminella) [inline spring-cloud-task-samples/taskprocessor/README.adoc]: We should mention in the doc the dependencies that this module has (redis in this case).

File: spring-cloud-task-samples/taskprocessor/src/main/java/io/spring/TaskProcessor.java [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+ * for a task and sends a {@link TaskLaunchRequest} message to a task sink.
+ *
+ * @author Glenn Renfro
+ */
+@EnableBinding(Processor.class)
+@EnableConfigurationProperties(TaskProcessorProperties.class)
+public class TaskProcessor {
+
+	@Autowired
+	TaskProcessorProperties processorProperties;
```
Comment (mminella) [inline spring-cloud-task-samples/taskprocessor/src/main/java/io/spring/TaskProcessor.java]: There a reason for the default scope?

File: spring-cloud-task-samples/taskprocessor/src/main/java/io/spring/TaskProcessor.java [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+
+	@Transformer(inputChannel = Processor.INPUT, outputChannel = Processor.OUTPUT)
+	public Object setupRequest(Message<?> message) {
```
Comment (mminella) [inline spring-cloud-task-samples/taskprocessor/src/main/java/io/spring/TaskProcessor.java]: Per @mbogoevici 's previous comment, are we updating this to take something other than a message and return the `TaskLaunchRequest`?

File: spring-cloud-task-samples/taskprocessor/src/main/java/io/spring/TaskProcessor.java [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+		Map<String, String> properties = new HashMap<String,String>();
+		properties.put("server.port", "0");
```
Comment (mminella) [inline spring-cloud-task-samples/taskprocessor/src/main/java/io/spring/TaskProcessor.java]: Why is this required?

File: spring-cloud-task-samples/taskprocessor/src/main/java/io/spring/TaskProcessor.java (hunk 8 above)
Comment (cppwfs) [inline spring-cloud-task-samples/taskprocessor/src/main/java/io/spring/TaskProcessor.java]: DOH!

File: spring-cloud-task-samples/taskprocessor/src/main/java/io/spring/TaskProcessor.java [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+		}
+		if(processorProperties.getDataSourceDriverClassName() != null){
+			properties.put("spring_datasource_driverClassName",processorProperties.getDataSourceDriverClassName());
+		}
+		if(processorProperties.getDataSourceUserName() != null){
+			properties.put("spring_datasource_username",processorProperties.getDataSourceUserName());
+		}
+		if(processorProperties.getDataSourcePassword() != null){
+			properties.put("spring_datasource_password",processorProperties.getDataSourcePassword());
+		}
```
Comment (mminella) [inline spring-cloud-task-samples/taskprocessor/src/main/java/io/spring/TaskProcessor.java]: For all of the above, use `StringUtils#hasText` since an empty string would cause errors.

File: spring-cloud-task-samples/taskprocessor/src/main/java/io/spring/TaskProcessorApplication.java [hunk 12]
```diff
@@ -0,0 +1,15 @@
+package io.spring;
```
Comment (mminella) [inline spring-cloud-task-samples/taskprocessor/src/main/java/io/spring/TaskProcessorApplication.java]: Missing license

Comment (cppwfs) [inline spring-cloud-task-samples/taskprocessor/src/main/java/io/spring/TaskProcessorApplication.java]: GAAK!

File: spring-cloud-task-samples/taskprocessor/src/main/java/io/spring/TaskProcessor.java (hunk 11 above)
Comment (cppwfs) [inline spring-cloud-task-samples/taskprocessor/src/main/java/io/spring/TaskProcessor.java]: good catch

File: spring-cloud-task-samples/taskprocessor/src/main/java/io/spring/TaskProcessorProperties.java [hunk 13]
```diff
@@ -0,0 +1,125 @@
...
+
+	public String getDataSourcePassword() {
+		return dataSourcePassword;
+	}
+
+	public void setDataSourcePassword(String dataSourcePassword) {
+		this.dataSourcePassword = dataSourcePassword;
+	}
+
+	private String dataSourcePassword;
```
Comment (mminella) [inline spring-cloud-task-samples/taskprocessor/src/main/java/io/spring/TaskProcessorProperties.java]: This should be moved to the top

File: spring-cloud-task-samples/taskprocessor/src/test/java/io/spring/TaskProcessorApplicationTests.java [hunk 14]
```diff
@@ -0,0 +1,48 @@
+package io.spring;
```
Comment (mminella) [inline spring-cloud-task-samples/taskprocessor/src/test/java/io/spring/TaskProcessorApplicationTests.java]: Missing license

File: spring-cloud-task-samples/taskprocessor/src/test/java/io/spring/TaskProcessorApplicationTests.java [hunk 15]
```diff
@@ -0,0 +1,48 @@
...
+	@Bindings(TaskProcessor.class)
+	protected Processor channels;
+
+	@Autowired
+	protected MessageCollector collector;
+
+		@Test
+		public void test() throws InterruptedException{
+			channels.input().send(new GenericMessage<Object>(DEFAULT_PAYLOAD));
+			Map properties = new HashMap<String, String>();
```
Comment (mminella) [inline spring-cloud-task-samples/taskprocessor/src/test/java/io/spring/TaskProcessorApplicationTests.java]: Any reason the generics on the `HashMap` instead of on the `Map` and having `<>` on `HashMap`?

File: spring-cloud-task-samples/tasksink/src/main/java/io/spring/TaskSinkApplication.java [hunk 16]
```diff
@@ -0,0 +1,19 @@
+package io.spring;
```
Comment (cppwfs) [inline spring-cloud-task-samples/tasksink/src/main/java/io/spring/TaskSinkApplication.java]: Missing license

File: spring-cloud-task-samples/tasksink/src/test/java/io/spring/TaskSinkApplicationTests.java [hunk 17]
```diff
@@ -0,0 +1,65 @@
+package io.spring;
```
Comment (cppwfs) [inline spring-cloud-task-samples/tasksink/src/test/java/io/spring/TaskSinkApplicationTests.java]: Missing license

File: spring-cloud-task-samples/tasksink/README.adoc [hunk 18]
```diff
@@ -0,0 +1,15 @@
...
+== Requirements:
+
+* Java 7 or Above
+
+== Build:
+
+[source,shell,indent=2]
+----
+$ ./mvnw clean install -DskipInstall=false
+----
```
Comment (mminella) [inline spring-cloud-task-samples/tasksink/README.adoc]: Can we specify dependencies as well as how to run this?

File: spring-cloud-task-stream/src/test/java/org/springframework/cloud/task/launcher/TaskLauncherSinkTests.java [hunk 19]
```diff
@@ -0,0 +1,86 @@
+package org.springframework.cloud.task.launcher;
```
Comment (cppwfs) [inline spring-cloud-task-stream/src/test/java/org/springframework/cloud/task/launcher/TaskLauncherSinkTests.java]: Missing license

File: spring-cloud-task-samples/tasksink/src/main/java/io/spring/TaskSinkApplication.java (hunk 16 above)
Comment (mminella) [inline spring-cloud-task-samples/tasksink/src/main/java/io/spring/TaskSinkApplication.java]: License

File: spring-cloud-task-samples/tasksink/src/test/java/io/spring/TaskSinkApplicationTests.java [hunk 20]
```diff
@@ -0,0 +1,65 @@
...
+	@After
+	public void tearDown(){
+		if(context != null){
+			context.close();
+		}
+	}
+
+	@Autowired
+	@Bindings(TaskLauncherSink.class)
+	private Sink sink;
```
Comment (mminella) [inline spring-cloud-task-samples/tasksink/src/test/java/io/spring/TaskSinkApplicationTests.java]: This belongs above the methods.

File: spring-cloud-task-samples/tasksink/src/test/java/io/spring/TaskSinkApplicationTests.java (hunk 17 above)
Comment (mminella) [inline spring-cloud-task-samples/tasksink/src/test/java/io/spring/TaskSinkApplicationTests.java]: License

File: spring-cloud-task-samples/tasksink/src/test/java/io/spring/TaskSinkApplicationTests.java [hunk 21]
```diff
@@ -0,0 +1,65 @@
...
+	public void testLaunch() {
+		context.register(TaskSinkApplication.class, TaskSinkConfiguration.class);
+		context.refresh();
+		this.sink = context.getBeansOfType(Sink.class).values().iterator().next();
+		assertNotNull(this.sink.input());
+
+		TaskSinkConfiguration.TestTaskLauncher testTaskLauncher =
+				(TaskSinkConfiguration.TestTaskLauncher) context.getBean(TaskSinkConfiguration.TestTaskLauncher.class);
+
+		Map<String, String> properties = new HashMap<String,String>();
```
Comment (mminella) [inline spring-cloud-task-samples/tasksink/src/test/java/io/spring/TaskSinkApplicationTests.java]: Diamond operator?

File: spring-cloud-task-samples/tasksink/src/test/java/io/spring/TaskSinkApplicationTests.java [hunk 22]
```diff
@@ -0,0 +1,65 @@
...
+	}
+
+	@Autowired
+	@Bindings(TaskLauncherSink.class)
+	private Sink sink;
+
+	@Test
+	public void testLaunch() {
+		context.register(TaskSinkApplication.class, TaskSinkConfiguration.class);
+		context.refresh();
```
Comment (mminella) [inline spring-cloud-task-samples/tasksink/src/test/java/io/spring/TaskSinkApplicationTests.java]: Is there a reason to do this within the test instead of using the Spring runner and `@ContextConfiguration`?

File: spring-cloud-task-stream/src/main/java/org/springframework/cloud/task/launcher/TaskLaunchRequest.java [hunk 23]
```diff
@@ -0,0 +1,116 @@
...
+	private String artifact;
+	private String taskGroupId;
+	private String taskVersion;
+	private String taskExtension;
+	private String taskClassifier;
+	private Map<String, String> properties;
+
+	public TaskLaunchRequest(String artifact, String taskGroupId, String taskVersion,
+							 String taskExtension, String taskClassifier,
+							 Map<String, String> properties) {
```
Comment (mminella) [inline spring-cloud-task-stream/src/main/java/org/springframework/cloud/task/launcher/TaskLaunchRequest.java]: Is it ok for any of these to be null/empty strings?

File: spring-cloud-task-stream/src/main/java/org/springframework/cloud/task/launcher/TaskLaunchRequest.java [hunk 24]
```diff
@@ -0,0 +1,116 @@
...
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Request that contains the resource and property information required by the
+ * TaskLauncherSink to launch the task.
+ *
+ * @author Glenn Renfro
+ */
+public class TaskLaunchRequest implements Serializable{
```
Comment (mminella) [inline spring-cloud-task-stream/src/main/java/org/springframework/cloud/task/launcher/TaskLaunchRequest.java]: `serialVersionUID`?

File: spring-cloud-task-stream/src/main/java/org/springframework/cloud/task/launcher/TaskLaunchRequest.java (hunk 23 above)
Comment (cppwfs) [inline spring-cloud-task-stream/src/main/java/org/springframework/cloud/task/launcher/TaskLaunchRequest.java]: taskClassifier and taskExtension.  The others need to be checked.

File: spring-cloud-task-stream/src/main/java/org/springframework/cloud/task/launcher/TaskLaunchRequest.java [hunk 25]
```diff
@@ -0,0 +1,116 @@
...
+		this.artifact = artifact;
+		this.taskGroupId = taskGroupId;
+		this.taskVersion = taskVersion;
+		this.taskExtension = taskExtension;
+		this.taskClassifier = taskClassifier;
+		this.properties = properties == null ? new HashMap<String,String>() :
+				properties;
+	}
+
+	public String getTaskGroupId() {
```
Comment (mminella) [inline spring-cloud-task-stream/src/main/java/org/springframework/cloud/task/launcher/TaskLaunchRequest.java]: Setter javadoc for this class.  I hope it's intuitive that these elements map to maven coordinates, but we should be explicit.

File: spring-cloud-task-stream/src/main/java/org/springframework/cloud/task/launcher/TaskLaunchRequest.java [hunk 26]
```diff
@@ -0,0 +1,116 @@
...
+
+	@Override
+	public String toString() {
+		return "TaskLaunchRequest{" +
+				"taskName='" + artifact + '\'' +
+				", taskGroupId='" + taskGroupId + '\'' +
+				", taskVersion='" + taskVersion + '\'' +
+				", taskExtension='" + taskExtension + '\'' +
+				", taskClassifier='" + taskClassifier + '\'' +
+				'}';
```
Comment (mminella) [inline spring-cloud-task-stream/src/main/java/org/springframework/cloud/task/launcher/TaskLaunchRequest.java]: Nitpick: Thoughts on generating the maven coordinates format for the properties that matter: `{taskName: foo, coordinates: io.spring:foo:1.0.0.BUILD-SNAPSHOT:exec:jar}`

File: spring-cloud-task-stream/src/main/java/org/springframework/cloud/task/launcher/TaskLaunchRequest.java [hunk 27]
```diff
@@ -0,0 +1,116 @@
...
+	}
+
+	@Override
+	public boolean equals(Object o) {
```
Comment (mminella) [inline spring-cloud-task-stream/src/main/java/org/springframework/cloud/task/launcher/TaskLaunchRequest.java]: If you override `.equals`, you typically want to override `.hashCode`...

File: spring-cloud-task-stream/src/main/java/org/springframework/cloud/task/launcher/TaskLaunchRequest.java [hunk 28]
```diff
@@ -0,0 +1,116 @@
...
+		if (this == o){
+			return true;
+		}
+		if (o == null || getClass() != o.getClass()){
+			return false;
+		}
+
+		TaskLaunchRequest that = (TaskLaunchRequest) o;
```
Comment (mminella) [inline spring-cloud-task-stream/src/main/java/org/springframework/cloud/task/launcher/TaskLaunchRequest.java]: Given the lack of validation in the constructor, you run the risk of NPEs in the if statements below...

File: spring-cloud-task-stream/src/main/java/org/springframework/cloud/task/launcher/TaskLaunchRequest.java [hunk 29]
```diff
@@ -0,0 +1,116 @@
...
+		if (!taskGroupId.equals(that.taskGroupId)){
+			return false;
+		}
+		if (!taskVersion.equals(that.taskVersion)){
+			return false;
+		}
+		if (!taskExtension.equals(that.taskExtension)){
+			return false;
+		}
+		if (taskClassifier != null ? !taskClassifier.equals(that.taskClassifier) : that.taskClassifier != null){
```
Comment (mminella) [inline spring-cloud-task-stream/src/main/java/org/springframework/cloud/task/launcher/TaskLaunchRequest.java]: I had to do a double take when reading this line...

File: spring-cloud-task-stream/src/main/java/org/springframework/cloud/task/launcher/TaskLauncherSink.java [hunk 30]
```diff
@@ -0,0 +1,70 @@
...
+import org.springframework.integration.annotation.ServiceActivator;
+import org.springframework.util.Assert;
+
+/**
+ * A sink stream application that launches a tasks.
+ *
+ * @author Glenn Renfro
+ */
+
+@EnableBinding(Sink.class)
```
Comment (mminella) [inline spring-cloud-task-stream/src/main/java/org/springframework/cloud/task/launcher/TaskLauncherSink.java]: Thoughts on not using the defaults for this to allow it to be combinable (I'm open to either option...looking for opinions)?

File: spring-cloud-task-stream/src/test/java/org/springframework/cloud/task/launcher/TaskLauncherSinkTests.java [hunk 31]
```diff
@@ -0,0 +1,86 @@
...
+	@After
+	public void tearDown(){
+		if(context != null){
+			context.close();
+		}
+	}
+
+	@Autowired
+	@Bindings(TaskLauncherSink.class)
+	private Sink sink;
```
Comment (mminella) [inline spring-cloud-task-stream/src/test/java/org/springframework/cloud/task/launcher/TaskLauncherSinkTests.java]: This should be above the setup/teardown

File: spring-cloud-task-stream/src/test/java/org/springframework/cloud/task/launcher/TaskLauncherSinkTests.java [hunk 32]
```diff
@@ -0,0 +1,86 @@
...
+
+	@Test
+	public void testSuccess() {
+		context.register(TaskLauncherSinkApplication.class, TaskConfiguration.class);
+		context.refresh();
```
Comment (mminella) [inline spring-cloud-task-stream/src/test/java/org/springframework/cloud/task/launcher/TaskLauncherSinkTests.java]: Any reason to do this in the tests instead of using SpringRunner and `@ConfigurationContext`?

File: spring-cloud-task-stream/src/main/java/org/springframework/cloud/task/launcher/TaskLauncherSink.java (hunk 30 above)
Comment (mbogoevici) [inline spring-cloud-task-stream/src/main/java/org/springframework/cloud/task/launcher/TaskLauncherSink.java]: Depends on the ultimate goal ... if this is to be deployed within SCDF then it certainly makes sense for it to be INPUT. Outside of that, I would doubt that the Task Launcher would meaningfully live within an application that also happens to process streaming messages, at least for a first pass. Should that be the case, I believe we could make the target interface configurable on the long run with a bit more postprocessing magic on `@EnableTaskLauncher`.

**Comment by mminella** [comment]
Squashed and merged as aad0a0b1ee2ffe8664aab43b69aa6686d2136bd1
